WordPress/src/main/java/org/wordpress/android/ui/media/services/MediaUploadService.java

```diff
@@ -193,21 +200,32 @@ private void uploadNextInQueue() {
         // somehow lost our reference to the site, complete this action
         if (site == null) {
             AppLog.i(T.MEDIA, "Unexpected state, site is null. Skipping this request - MediaUploadService.");
-            stopServiceIfUploadsComplete();
+            updatePostAndStopServiceIfUploadsComplete();
             return;
         }
 
         dispatchUploadAction(next, site);
     }
 
-    private void completeUploadWithId(int id) {
+    private synchronized void completeUploadWithId(int id) {
         MediaModel media = getMediaFromQueueById(id);
         if (media != null) {
             mInProgressUploads.remove(media);
             trackUploadMediaEvents(AnalyticsTracker.Stat.MEDIA_UPLOAD_STARTED, media, null);
         }
     }
 
+    // this keeps a map for all completed media for each post, so we can process the post easily
+    // in one go later
+    private void addMediaToPostCompletedMediaListMap(MediaModel media) {
+        List<MediaModel> mediaListForPost = mCompletedUploads.get(media.getLocalPostId());
+        if (mediaListForPost == null) {
+            mediaListForPost = new ArrayList<>();
+        }
+        mediaListForPost.add(media);
+        mCompletedUploads.put(new Integer(media.getLocalPostId()), mediaListForPost);
+    }
+
     private MediaModel getMediaFromQueueById(int id) {
```

Comment:
In the alternate-universe version of this PR (https://github.com/wordpress-mobile/WordPress-Android/pull/5991) you renamed this method to `getMediaFromInProgressQueueById()`, and I think that was a good change - want to re-add it in this PR?